jax/experimental/shard_map.py

```diff
@@ -0,0 +1,912 @@
+                           tree_structure, tree_leaves)
+from jax._src.tree_util import (broadcast_prefix, prefix_errors, PyTreeDef,
+                                _generate_key_paths, KeyPath)
+
+P = PartitionSpec
+
+map, unsafe_map = util.safe_map, map
+zip, unsafe_zip = util.safe_zip, zip
+traceback_util.register_exclusion(__file__)
+
+# API
+
+Specs = Any  # PyTree[PartitionSpec]
+
+@traceback_util.api_boundary
+def shard_map(f: Callable, mesh: Mesh, in_specs: Specs, out_specs: Specs,
+              check_rep: bool = True):
+  if not callable(f):
+    raise TypeError("shard_map requires a callable for its first argument, "
+                    f"but got {f} of type {type(f)}.")
+  if not isinstance(mesh, Mesh):
+    raise TypeError("shard_map requires a `jax.sharding.Mesh` instance for its "
+                    f"second argument, but got {mesh} of type {type(mesh)}.")
+  _check_specs(SpecErrorType.input, in_specs)
+  _check_specs(SpecErrorType.out, out_specs)
+
+  @traceback_util.api_boundary
+  def wrapped(*args):
+    fun = lu.wrap_init(f)
+    args_flat, in_tree = tree_flatten(args)
+    try: in_specs_flat = broadcast_prefix(in_specs, args)
+    except ValueError:
+      e, *_ = prefix_errors(in_specs, args)
+      raise e('shard_map in_specs') from None
+    _check_specs_vs_args(f, mesh, in_tree, in_specs, in_specs_flat, args_flat)
+    in_names_flat = tuple(map(_canonicalize_spec, in_specs_flat))
+    flat_fun, out_tree = flatten_fun_nokwargs(fun, in_tree)
+
+    @memoize
+    def out_names_thunk():
+      dummy = tree_unflatten(out_tree(), [object()] * out_tree().num_leaves)
+      try: out_specs_flat = broadcast_prefix(out_specs, dummy)
+      except ValueError:
+        e, *_ = prefix_errors(out_specs, dummy)
+        raise e('shard_map out_specs') from None
+      return tuple(map(_canonicalize_spec, out_specs_flat))
+    try:
+      out_flat = shard_map_p.bind(
+          flat_fun, *args_flat, mesh=mesh, in_names=in_names_flat,
+          out_names_thunk=out_names_thunk, check_rep=check_rep)
+    except _SpecError as e:
+      fails, = e.args
+      msg = _spec_rank_error(SpecErrorType.out, f, out_tree(), out_specs, fails)
+      if any(fail is not no_fail and not fail.shape for fail in fails):
+        msg += (" In particular, for rank 0 outputs which are not constant "
+                "over the mesh, add at least one (singleton) axis to them so "
+                "that they can be concatenated using out_specs.")
+      raise ValueError(msg) from None
+    except _RepError as e:
+      fails, = e.args
+      msg = _rep_error(f, mesh, out_tree(), out_specs, fails)
+      raise ValueError(msg) from None
+    return tree_unflatten(out_tree(), out_flat)
+  return wrapped
+
+# Internally use AxisNames = Dict[int, Tuple[AxisName, ...]], not PartitionSpecs
+AxisName = Hashable
+AxisNames = Dict[int, Tuple[AxisName, ...]]  # TODO(mattjj): make it hashable
+def _canonicalize_spec(spec: PartitionSpec) -> AxisNames:
```

Comment:
Isn't this pjit.array_mapping?

Reply:
https://github.com/google/jax/blob/main/jax/_src/pjit.py#L1929 -- yes it is! Can we reuse?

Reply:
It looks different to me. The output of `_canonicalize_spec` is a `Dict[int, Tuple[AxisName, ...]]`, but the output of the `get_array_mapping` you linked is an ordered version of `Dict[AxisName, int]` (unioned with some other stuff). That is, I think they're perhaps flipped?

Reply:
Ohh yeah, I misread. Cool!